### PR TITLE
dnsdist: Add a new chain of rules triggered after cache insertion

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

The general idea is to be able to store the unedited version into the cache while delivering a different version to the actual client. This is useful when one is sending different answers to different clients, like when dealing with abuse traffic, but still want to be able to cache the initial response from the backend. We already have a chain of rules that are triggered after a cache-hit, but until now we lacked the ability to trigger after getting the response corresponding to a cache-miss.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [x] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Comment by rgacogne** [comment]
CI failures are unrelated.

File: pdns/dnsdistdist/docs/reference/config.rst [hunk 1]
```diff
@@ -1103,6 +1103,14 @@ Status, Statistics and More
 
   :param int top: How many response rules to return.
 
+.. function:: getTopCacheInsertedResponseRules([top])
+
+  .. versionadded:: 1.8.0
+
+  Return the cache-inserted response rules that matched the most.
+
+  :param int top: How many response rules to return.
```
Comment (omoerbeek) [inline pdns/dnsdistdist/docs/reference/config.rst:1112]: Default value of `top` missing

File: pdns/dnsdistdist/docs/reference/config.rst [hunk 2]
```diff
@@ -1260,6 +1268,19 @@ Status, Statistics and More
 
   * ``showUUIDs=false``: bool - Whether to display the UUIDs, defaults to false.
 
+.. function:: topCacheInsertedResponseRules([top [, options]])
+
+  .. versionadded:: 1.8.0
+
+  This function shows the cache-inserted response rules that matched the most.
+
+  :param int top: How many rules to show.
```
Comment (omoerbeek) [inline pdns/dnsdistdist/docs/reference/config.rst:1277]: Default value  of `top` missing

File: pdns/dnsdist-lua-rules.cc [hunk 3]
```diff
@@ -333,18 +349,30 @@ void setupLuaRules(LuaContext& luaCtx)
     return rulesToString(getTopRules(*rules, top.get_value_or(10)), vars);
   });
 
-  luaCtx.writeFunction("getCacheHitResponseRules", [](boost::optional<unsigned int> top) {
```
[the remark this answers is not in the document]
Comment (Habbie) [reply pdns/dnsdist-lua-rules.cc:336]: Ah! I see the 'new' names in the old docs now. Great :)

**Review by Habbie — APPROVED** [review]

**Pull request merged**